janus_server/src/aggregator.rs

```diff
@@ -194,10 +217,119 @@ fn with_cloned_value<T: Clone + Sync + Send>(
+    fn type_uri(&self) -> &'static str {
+        match self {
+            PpmProblemType::UnrecognizedMessage => "urn:ietf:params:ppm:error:unrecognizedMessage",
+            PpmProblemType::UnrecognizedTask => "urn:ietf:params:ppm:error:unrecognizedTask",
+            PpmProblemType::OutdatedConfig => "urn:ietf:params:ppm:error:outdatedConfig",
+            PpmProblemType::StaleReport => "urn:ietf:params:ppm:error:staleReport",
+            PpmProblemType::InvalidHmac => "urn:ietf:params:ppm:error:invalidHmac",
+        }
+    }
+
+    /// Returns a human-readable summary of a problem type.
+    fn description(&self) -> &'static str {
+        match self {
+            PpmProblemType::UnrecognizedMessage => {
+                "The message type for a response was incorrect or the payload was malformed"
+            }
+            PpmProblemType::UnrecognizedTask => {
+                "An endpoint received a message with an unknown task ID."
+            }
+            PpmProblemType::OutdatedConfig => {
+                "The message was generated using an outdated configuration."
+            }
+            PpmProblemType::StaleReport => {
+                "Report could not be processed because it arrived too late."
+            }
+            PpmProblemType::InvalidHmac => "The aggregate message's HMAC was not valid.",
+        }
+    }
+}
+
+/// The media type for problem details formatted as a JSON document, per RFC 7807.
+static PROBLEM_DETAILS_JSON_MEDIA_TYPE: &str = "application/problem+json";
+
+/// Construct an error response in accordance with §3.1.
+/// TODO (PR abetterinternet/ppm-specification#208): base64-encoding the TaskID has not yet been
+/// adopted in the specification, and may change.
+/// TODO (issue abetterinternet/ppm-specification#209): The handling of the instance, title,
+/// detail, and taskid fields are subject to change.
+fn build_problem_details_response(
+    error_type: PpmProblemType,
+    task_id: TaskId,
+    endpoint: &Url,
+) -> Response {
+    // So far, 400 Bad Request seems to be the appropriate choice for each defined problem type.
+    let status = StatusCode::BAD_REQUEST;
+    warp::reply::with_status(
+        warp::reply::with_header(
+            warp::reply::json(&serde_json::json!({
+                "type": error_type.type_uri(),
+                "title": error_type.description(),
+                "status": status.as_u16(),
+                "detail": error_type.description(),
+                "instance": endpoint.as_str(),
+                "taskid": base64::encode(task_id.as_bytes()),
+            })),
+            http::header::CONTENT_TYPE,
+            PROBLEM_DETAILS_JSON_MEDIA_TYPE,
+        ),
+        status,
+    )
+    .into_response()
+}
+
+/// Produces a closure that will transform applicable errors into a problem details JSON object.
+/// (See RFC 7807) The returned closure is meant to be used in a warp `map` filter.
+fn error_handler<C, R>(
+    aggregator: &Aggregator<C>,
```

Comment:
nit: all you need here is the aggregator's endpoint, so you could remove some generic complexity by doing `fn error_handler(endpoint: Url) -> ...`